ballista/rust/executor/src/executor.rs

```diff
@@ -110,15 +121,224 @@ impl Executor {
             ))
         }?;
 
-        let partitions = exec.execute_shuffle_write(part, task_ctx).await?;
+        let (task, abort_handle) =
+            futures::future::abortable(exec.execute_shuffle_write(part, task_ctx));
+
+        {
+            let mut abort_handles = self.abort_handles.lock().await;
+            abort_handles.insert(
+                PartitionId {
+                    job_id: job_id.clone(),
+                    stage_id,
+                    partition_id: part,
+                },
+                abort_handle,
+            );
+        }
+
+        let partitions = task.await??;
```

Comment:
two ??

Reply:
Yeah, when you wrap the inner future in an `Abortable` it adds another `Result` layer (to indicate whether the task was cancelled). So the first `?` is unwrapping the cancellation result (i.e. `Err` if the task was cancelled) and the second `?` is unwrapping the inner future's result.